dicts/dicts.gobra

```diff
@@ -0,0 +1,466 @@
+    - https://www.why3.org/stdlib/map.html
+*/
+
+// This package defines lemmas for mathematical maps (aka dictionaries) commonly used
+// in specifications.
+package dicts
+
+import "util"
+import "sets"
+
+// A dictionary is empty if its domain is empty.
+ghost
+decreases
+pure func IsEmpty(d dict[int]int) bool {
+	return sets.IsEmpty(domain(d))
+}
+
+// Returns the empty dictionary.
+ghost
+ensures IsEmpty(result)
+decreases
+pure func Empty() (result dict[int]int) {
+	return dict[int]int{}
+}
+
+// Retrieves the value associated with the key, if present, as an option.
+ghost
+decreases
+pure func Get(d dict[int]int, k int) option[int] {
+	return k in domain(d) ? some(d[k]) : none[int]
+}
+
+// Keep all key-value pairs corresponding to the set of keys provided.
+ghost
+ensures forall x int :: {result[x]} (x in domain(d) && x in xs) ==>
+	(x in domain(result) && result[x] == d[x])
+ensures forall x int :: {x in domain(result)} x in domain(result) ==>
+	(x in domain(d) && x in xs)
+ensures domain(result) == xs intersection domain(d)
+opaque
+decreases xs
+pure func Restrict(d dict[int]int, xs set[int]) (result dict[int]int) {
+	return let ys := (xs intersection domain(d)) in
+		(sets.IsEmpty(ys) ? Empty() :
+			(let y := sets.Choose(ys) in
+			(let yr := sets.Remove(ys, y) in
+			(let _ := sets.IntersectLenUpper(xs, domain(d)) in
+			 Restrict(d, yr)[y = d[y]]))))
+}
+
+// Remove all key-value pairs corresponding to the set of keys provided.
+ghost
+decreases
+pure func RemoveKeys(d dict[int]int, xs set[int]) (result dict[int]int) {
+	return Restrict(d, domain(d) setminus xs)
+}
+
+// Remove a key-value pair. Returns unmodified dictionary if they key is not found.
+ghost
+ensures len(result) <= len(d)
+ensures k in domain(d) ==> len(result) == len(d) - 1
+ensures !(k in domain(d)) ==> len(result) == len(d)
+decreases
+pure func Remove(d dict[int]int, k int) (result dict[int]int) {
+	return let ys := (sets.Remove(domain(d), k)) in
+		let _ := util.Asserting(ys intersection domain(d) == ys) in
+		RemoveKeys(d, sets.SingletonSet(k))
+}
+
+// True iff k maps to the same value or is not in the domains of d1 and d2.
+ghost
+decreases
+pure func IsEqualOnKey(d1, d2 dict[int]int, k int) bool {
+	return !(k in domain(d1) || k in domain(d2)) ||
+		(k in domain(d1) && k in domain(d2) && d1[k] == d2[k])
+}
+
+// True iff if d1 is a subset of d2.
+ghost
+decreases
+pure func IsSubset(d1, d2 dict[int]int) bool {
+	return domain(d1) subset domain(d2) &&
+		forall k int :: {IsEqualOnKey(d1, d2, k)} {k in domain(d1)} (k in domain(d1)) ==>
+		IsEqualOnKey(d1, d2, k)
+}
+
+// Union of two dictionaries. Does not require disjoint domains: on the intersection,
+// values from the second dictionary are chosen.
+ghost
+ensures domain(result) == domain(d1) union domain(d2)
+ensures forall k int :: {result[k]} (k in domain(d2)) ==> result[k] == d2[k]
+ensures forall k int :: {result[k]}  (!(k in domain(d2)) && k in domain(d1)) ==>
+	result[k] == d1[k]
+opaque
```

Comment:
To me, `opaque` should always come after `ghost`. We should have a consistent order of annotations to make it easy to spot things quickly